cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2016 SnappyData, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+package org.apache.spark.memory
+
+import scala.collection.mutable.ArrayBuffer
+
+import com.gemstone.gemfire.internal.cache.GemFireCacheImpl
+import com.pivotal.gemfirexd.internal.engine.Misc
+import org.apache.spark.SparkConf
+import scala.collection.mutable
+
+import org.apache.spark.storage.{BlockStatus, BlockId}
+
+/**
+  * When there is request for execution or storage memory, critical up and eviction up
+  * events are checked. If they are set, try to free the memory cached by Spark rdds
+  * by calling memoryStore.evictBlocksToFreeSpace. If enough memory cannot be freed,
+  * return the call and let Spark take a corrective action.
+  * In such cases Spark either fails the task or move the current RDDs data to disk.
+  * If the critical and eviction events are not set, it asks the UnifiedMemoryManager
+  * to allocate the space.
+  * @param conf
+  * @param maxMemory
+  * @param numCores
+  */
+class SnappyUnifiedMemoryManager private[memory](
+    conf: SparkConf,
+    override val maxMemory: Long,
+    numCores: Int) extends UnifiedMemoryManager(conf,
+  maxMemory,
+  (maxMemory * conf.getDouble("spark.memory.storageFraction", 0.5)).toLong,
+  numCores) {
+
+  def this(conf: SparkConf, numCores: Int) = {
+    this( conf,
+      SnappyUnifiedMemoryManager.getMaxMemory(conf),
+      numCores)
+  }
+  private val listener = Misc.getMemStoreBooting.thresholdListener()
+
+  private def freeMemory(blockId: Option[BlockId], numBytes: Long,
+      evictedBlocks: mutable.Buffer[(BlockId, BlockStatus)]): Boolean = {
+    val freeMemory = Runtime.getRuntime.freeMemory()
+    if (freeMemory < numBytes) {
+
+      storageMemoryPool.memoryStore.evictBlocksToFreeSpace(blockId,
+        numBytes - freeMemory,
+        evictedBlocks) || !listener.isCritical
```

Comment:
> Ok, I thought at least for storage memory, job will not fail if enough memory is not acquired in acquireStorageMemory(). It will just move the RDD to disk (if possible) or recompute RDD on the fly when needed.

I think in the new model if storage pool fails to acquire the required blocks, then there are no more blocks available in Spark layer (since it can steal from execution pool).

@hbhanawat Assuming that to be the case, I think Shirish is right that it should better start dropping cached blocks in that case for storage blocks. As per the default setting of maxMemory==GemXD_eviction below, similar is the effect that beyond eviction limit start dropping storage blocks (otherwise you should also change maxMemory below to criticalLimit to keep it consistent with the reasoning here). So a flag can be passed to freeMemory for the two cases:

- if freeMemory() has been invoked from acquireStorageMemory, return false if evictBlocksToFreeSpace fails and evictionUp is still true
- if freeMemory() has been invoked from acquireExecutionMemory then only fail if criticalUp is still true after evictBlocksToFreeSpace

Secondly should we allow it to fail the Spark way or should we fail by explicitly throwing LowMemoryException. Issue is that if the failure comes from GemXD layer it will be a LowMemoryException, but from Spark layer it will be like "unable to acquire ..." error. Perhaps we should somehow have Spark layer throw LowMemoryException after it has done retries and all instead of "unable to acquire ..." for consistency (message can still be same as "unable to acquire xxx bytes")